### PR TITLE
Expose interface to update ENR according to subscribed attestation subnets

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
Expose an interface to update ENR of DiscoveryNetwork according to long term subscribed attestation subnets.
https://pegasys1.atlassian.net/secure/RapidBoard.jspa?rapidView=67&modal=detail&selectedIssue=BC-286

### CODE REVIEW

File: util/src/main/java/tech/pegasys/artemis/util/config/Constants.java [hunk 1]
```diff
@@ -147,6 +147,15 @@
   public static final long FORK_RETRY_DELAY_SECONDS = 10; // in sec
   public static final long FORK_REFRESH_TIME_SECONDS = TimeUnit.MINUTES.toSeconds(5); // in sec
 
+  // Networking
+  public static final int GOSSIP_MAX_SIZE = 1048576; // bytes
```
Comment (mbaxter) [inline util/src/main/java/tech/pegasys/artemis/util/config/Constants.java:153]: Some of these constants are defined closer to the classes that use them - so you should make sure to deduplicate.

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/network/P2PNetwork.java [hunk 2]
```diff
@@ -88,6 +89,15 @@
...
   Optional<String> getEnr();
 
+  /**
+   * Updates the value of or creates a field for the Ethereum Node Record (ENR) for the local node,
+   * if one exists.
+   *
+   * @param fieldName
+   * @param value
+   */
+  void updateCustomENRField(String fieldName, Bytes value);
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/network/P2PNetwork.java]: (nit) Maybe just:
```suggestion
  void setENRField(String fieldName, Bytes value);
```

Comment (cemozerr) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/network/P2PNetwork.java]: updateCustomENRField is not the best, but there can be multiple different custom fields, so set wouldn't be too correct either. I'm undecided.

File: util/src/main/java/tech/pegasys/artemis/util/config/Constants.java (hunk 1 above)
Comment (cemozerr) [reply util/src/main/java/tech/pegasys/artemis/util/config/Constants.java:153]: done.

File: networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptions.java [hunk 3]
```diff
@@ -38,6 +47,7 @@ public synchronized void subscribeToCommittee(
     final UnsignedLong currentUnsubscribeSlot =
         unsubscriptionSlotByCommittee.getOrDefault(committeeIndex, ZERO);
     unsubscriptionSlotByCommittee.put(committeeIndex, max(currentUnsubscribeSlot, aggregationSlot));
+    handleCommitteeSubscriptionENRChange(committeeIndex);
```
Comment (ajsutton) [inline networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptions.java]: The ENR should only report the subnets that we randomly selected to subscribe to long term, not the ones that we temporarily subscribe to as part of our aggregation duties:

https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/validator.md#phase-0-attestation-subnet-stability:
>Maintain advertisement of the randomly selected subnets in their node's ENR attnets entry by setting the randomly selected subnet_id bits to True (e.g. ENR["attnets"][subnet_id] = True) for all persistent attestation subnets

Right now I think that's always an empty set because we aren't selecting long term subnets.

File: networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/core/RpcTimeouts.java [hunk 4]
```diff
@@ -23,9 +24,9 @@
 public abstract class RpcTimeouts {
 
   // The maximum time to wait for first byte of request response (time-to-first-byte).
-  static final Duration TTFB_TIMEOUT = Duration.ofSeconds(5);
+  static final Duration TTFB_TIMEOUT = Duration.ofSeconds(Constants.TTFB_TIMEOUT);
   // The maximum time for complete response transfer.
-  static final Duration RESP_TIMEOUT = Duration.ofSeconds(10);
+  static final Duration RESP_TIMEOUT = Duration.ofSeconds(Constants.RESP_TIMEOUT);
```
Comment (ajsutton) [inline networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/core/RpcTimeouts.java:29]: We're going to have to have a conversation about shoving every constant into one file at some point - it's getting very crowded in there....

File: networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/Eth2Network.java [hunk 5]
```diff
@@ -21,4 +22,6 @@
   void subscribeToAttestationCommitteeTopic(int committeeIndex);
 
   void unsubscribeFromAttestationCommitteeTopic(int committeeIndex);
+
+  void updateAttestationSubnetENRField(Bytes value);
```
Comment (ajsutton) [inline networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/Eth2Network.java]: I think we've wound up mixing committeeIndex and attestation subnet in this interface with the subscribe/unsubscribe talking in terms of committee index but the attestation subnet defining subnet indices. Prior to v0.11 that was a 1-1 mapping but now that the topic index is `committeeIndex % ATTESTATION_SUBNET_COUNT` they wind up different and something has to handle cases where two committees map to the same subnet (especially when we unsubscribe from one but not the other).

I'd be tempted to define this entirely in terms of subnet index so the `AttestationTopicSubscriptions` class is responsible for the mapping from committee index to subnet index and working out which subnets we should be subscribed from that.

But I'd also be very tempted to change `updateAttestationSubnetENRField` to be more discovery system agnostic so it's just `setLongTermAttestationSubnetSubscriptions(List<Integer> subnetIndices)` and that flows all the way down to `DiscoveryNetwork` which is responsible for converting the indices to the right format to put in an ENR.  Outside of `DiscoveryNetwork` shouldn't need to know how it's represented (or really even that it goes in the ENR).  In the future it might use discv5 topic searches rather than an enr field and ideally only `DiscoveryNetwork` would have to change.

File: networking/eth2/src/test/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptionsTest.java [hunk 6]
```diff
@@ -82,12 +85,67 @@ public void shouldPreserveLaterSubscriptionPeriodWhenEarlierSlotAdded() {
...
+    bitvector.setBit(4);
+
+    verify(eth2Network, times(1)).updateAttestationSubnetENRField(bitvector.serialize());
+
+    subscriptions.onSlot(slot1.plus(ONE));
+    verify(eth2Network, times(1)).updateAttestationSubnetENRField(bitvector.serialize());
+
+    bitvector.clearBit(4);
+    subscriptions.onSlot(slot2.plus(ONE));
+    verify(eth2Network).updateAttestationSubnetENRField(bitvector.serialize());
```
Comment (ajsutton) [inline networking/eth2/src/test/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptionsTest.java]: We shouldn't be calling `unsubscribeFromCommittee` when the subnet overlaps with one we stay subscribed to either.

File: networking/eth2/src/test/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptionsTest.java [hunk 7]
```diff
@@ -82,12 +85,67 @@ public void shouldPreserveLaterSubscriptionPeriodWhenEarlierSlotAdded() {
...
+    subscriptions.subscribeToCommittee(committeeIndex1, slot1);
+    subscriptions.subscribeToCommittee(committeeIndex2, slot2);
+
+    Bitvector bitvector = new Bitvector(ATTESTATION_SUBNET_COUNT);
+    bitvector.setBit(4);
+
+    verify(eth2Network, times(1)).updateAttestationSubnetENRField(bitvector.serialize());
+
+    subscriptions.onSlot(slot1.plus(ONE));
+    verify(eth2Network, times(1)).updateAttestationSubnetENRField(bitvector.serialize());
```
Comment (ajsutton) [inline networking/eth2/src/test/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptionsTest.java]: I suspect this should be
```suggestion
    verify(eth2Network, times(1)).updateAttestationSubnetENRField(any());
```

Otherwise the test will pass if the ENR field is updated with a different value because it's only asserting that particular value is only set once.

Comment (cemozerr) [reply networking/eth2/src/test/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptionsTest.java]: Yeah. Good catch.

File: networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/core/RpcTimeouts.java (hunk 4 above)
Comment (cemozerr) [reply networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/core/RpcTimeouts.java:29]: I understand, but I'm trying to be as consistent as possible. If we did decide to move the constants out of that file to individual files, I would not shove them there.

File: networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptions.java (hunk 3 above)
Comment (cemozerr) [reply networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptions.java]: My thinking was that, since we don't have the random selection of persistent subnets yet, for now, the new ENR attestation subnet bitfield functionality would be used to display attestation subnets we're subscribing for aggregation. That was probably an assumption I should have made more explicitly, and maybe not the best assumption to make. I will refactor this PR to explicitly expose an interface to subscribe and unsubscribe from randomly selected persistent attestation subnets.

File: networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/Eth2Network.java (hunk 5 above)
Comment (cemozerr) [reply networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/Eth2Network.java]: changing `updateAttestationSubnetENRField` to `setLongTermAttestationSubnetSubscriptions(List<Integer> subnetIndices)` makes a lot of sense to me.

Comment (cemozerr) [reply networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/Eth2Network.java]: However, the fact that DiscoveryNetwork is an instance of the P2PNetwork makes it impossible, as far as I understand, to keep any Discovery specific implementation detail out of P2PNetwork, DelegatingP2PNetwork, LibP2PNetwork, MockP2PNetwork etc. Any DiscoveryNetwork specific API I need to implement, I need to implement for all these classes as well.

Comment (ajsutton) [reply networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/Eth2Network.java]: I think all the discovery specific stuff should be *inside* DiscoveryNetwork, not exposed by it's API.  Just like the discovery library dependency is entire inside DiscoveryNetwork.

Comment (cemozerr) [reply networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/Eth2Network.java]: Makes total sense to me as discussed offline.

File: networking/eth2/src/test/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptionsTest.java [hunk 8]
```diff
@@ -82,7 +82,6 @@ public void shouldPreserveLaterSubscriptionPeriodWhenEarlierSlotAdded() {
     final UnsignedLong secondSlot = UnsignedLong.valueOf(15);
 
     subscriptions.subscribeToCommittee(committeeIndex, secondSlot);
-    subscriptions.subscribeToCommittee(committeeIndex, firstSlot);
```
Comment (mbaxter) [inline networking/eth2/src/test/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptionsTest.java]: Why was this line removed?

File: ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java [hunk 9]
```diff
@@ -40,6 +42,26 @@ public Bitvector(Bitvector bitvector) {
...
 
+  public Bitvector(List<Integer> indicesToSet, int size) {
+    this.size = size;
+    this.data = new BitSet(size);
+    if (indicesToSet.size() == 0) {
+      return;
+    }
+    checkArgument(
+        indicesToSet.stream().distinct().count() != indicesToSet.size(),
+        "Bitvector: Bit indices to set are not distinct");
```
Comment (mbaxter) [inline ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java]: Instead of adding a check here, you could just accept a `Set` of indices:
```
public Bitvector(Set<Integer> indicesToSet, int size) {
```

File: ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java [hunk 10]
```diff
@@ -40,6 +42,26 @@ public Bitvector(Bitvector bitvector) {
...
+    checkArgument(
+        indicesToSet.stream().min(Integer::compareTo).get() > -1,
+        "Bitvector: Bit index less than 0");
+    checkArgument(
+        indicesToSet.stream().max(Integer::compareTo).get() > size,
```
Comment (mbaxter) [inline ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java]:
```suggestion
        indicesToSet.stream().max(Integer::compareTo).get() >= size,
```

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/DiscoveryNetwork.java [hunk 11]
```diff
@@ -119,6 +124,12 @@ public void addStaticPeer(final String peerAddress) {
     return discoveryService.getEnr();
   }
 
+  public void setLongTermAttestationSubnetSubscriptions(List<Integer> subnetIds) {
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/DiscoveryNetwork.java]: What about:
```suggestion
  public void advertiseLongTermAttestationSubnetSubscriptions(final Set<Integer> subnetIds) {
```

Since the discovery network's role is limited to advertisements.

File: ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java (hunk 9 above)
Comment (ajsutton) [reply ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java]: That's quite an expensive check to add as well and I'm not sure it actually matters.  It's like creating a `Set` from a `List` - duplicates should just be ignored.

File: ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java (hunk 10 above)
Comment (ajsutton) [reply ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java]: This check is requiring us to iterate through the `List` a second time.  I'd suggest the constructor would just be:

```
public Bitvector(Iterable<Integer> indicesToSet, int size) {
  this(size);
  for (int i : indicesToSet) {
    setBit(i);
  }
}
```
The element index is then check as part of setting each bit so we only iterate once and any checks added in the future are automatically applied.  It basically just becomes syntactic sugar to make converting a list of ints to a Bitvector easy but does exactly the same thing as if you'd use the public APIs yourself.

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/DiscoveryNetwork.java (hunk 11 above)
Comment (ajsutton) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/DiscoveryNetwork.java]: Following on from Meredith's suggestion to use a `Set` in the `Bitvector` constructor, we should probably use a `Collection` here to allow callers to choose either a `List` or `Set`.  The order doesn't matter to us so being flexible is quite useful.

And when we add the piece to call this building up the indices as a `Set` would make a lot of sense.

**Review by ajsutton — APPROVED** [review]
LGTM. Would definitely be good to simplify the `Bitvector` constructor though.

File: networking/eth2/src/test/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptionsTest.java (hunk 8 above)
Comment (cemozerr) [reply networking/eth2/src/test/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationTopicSubscriptionsTest.java]: My bad. Thanks for the catch.

File: ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java (hunk 10 above)
Comment (cemozerr) [reply ssz/src/main/java/tech/pegasys/artemis/ssz/SSZTypes/Bitvector.java]: Makes a lot of sense. Done.

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/DiscoveryNetwork.java (hunk 11 above)
Comment (cemozerr) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/DiscoveryNetwork.java]: Makes sense.
